### PR TITLE
WIP: Reproducibility note

### PR DESCRIPTION
**Pull request by themightyoarfish**
This adds a Note on making experiments reproducible. 

It also adds Instructions for building the Documentation to `README.md`. Please ping if I missed any requirements.

I'm not sure what to do about the submodule changes. Please advise.

### CODE REVIEW

**Comment by t-vi** [comment]
I think there are known reproducibility issues with distributed (search the forums).
Also things like index_add (using atomic add) are not deterministic, this includes CTC loss - although the latter might possibly be fixable by fixing the precision of the atomic add operands.

**Comment by themightyoarfish** [comment]
Ok then the note should be amended, but would you agree that there should be some documentation on it? At least on a single machine this would do the trick, and finding this information was really hard.

**Comment by SsnL** [comment]
@themightyoarfish Please revert the submodule change. Moreover, it might be good to mention that reproducibility is not guaranteed across versions and/or architectures. E.g., a RNG on CUDA definitely behaves different than the RNG on CPU even though they use same seed; and when we optimize a sampling kernel in a release, the generated numbers might be different (although distribution is the same).

File: docs/source/notes/randomness.rst [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+However, in order to make computations deterministic on your specific problem on
+one specific platform and PyTorch release, there are a couple of steps to take.
+
+There are several pseudorandom number generators involved in various layers of
+the library, which you will need to seed manually to make runs reproducible.
+
+Numpy
+.....
+The numpy rng can be seeded with::
```
Comment (ezyang) [inline docs/source/notes/randomness.rst]: I wouldn't put the bit about Numpy here; not everyone is going to use Numpy. I'd put it after PyTorch, and preface it with, "if you, or any of the libraries you are using, use numpy..."

File: docs/source/notes/randomness.rst [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+PyTorch
+.......
+When running on the CPU, you should seed the RNG with::
+
+    import torch
+    torch.manual_seed(0)
+
+While this will seed all CUDA RNGs as well, you can do so explicitly with
+:meth:`torch.cuda.manual_seed` for the current device, or
+:meth:`torch.cuda.manual_seed_all` to seed the RNG for all GPUs
```
Comment (ezyang) [inline docs/source/notes/randomness.rst]: It's worth noting why you might *not* want to `manual_seed_all`, which is that the time it takes scales with the number of GPUs you seed, so if you have a lot of GPUs but are only using one of them, `manual_seed_all` will needlessly bloat your load time,

Comment (soumith) [reply docs/source/notes/randomness.rst]: `torch.manual_seed` will set all devices lazily, including all CUDA GPUs.
It USED to be that the behavior of `torch.manual_seed` only seeded CPU.
Hence, this section can simply be said: `To seed all devices with a fixed RNG, set `torch.manual_seed(0)`.

In fact, @ezyang wrote that patch :)

File: docs/source/notes/randomness.rst (hunk 1 above)
Comment (themightyoarfish) [reply docs/source/notes/randomness.rst]: Good suggestion; I just searched for `numpy` in the repo and deduced from that that numpy is used somewhere within PyTorch.

File: README.md [hunk 3]
```diff
@@ -224,6 +225,17 @@ Please note that PyTorch uses shared memory to share data between processes, so
 for multithreaded data loaders) the default shared memory segment size that container runs with is not enough, and you
 should increase shared memory size either with `--ipc=host` or `--shm-size` command line options to `nvidia-docker run`.
 
+### Building the Documentation
+
+To build documentation in various formats, you will need Sphinx.
+
+```
+conda install sphinx sphinx_rtd_theme
```
Comment (soumith) [inline README.md]: this should be:

```
cd docs/
pip install -r requirements.txt
```

Comment (themightyoarfish) [reply README.md]: Huh, totally did not see that that exists. Fixing.

**Comment by themightyoarfish** [comment]
Seems like we have a typo in the note
`Furthermore, results need to be
reproducible between CPU and GPU executions`, there's a `not` missing here.

**Comment by themightyoarfish** [comment]
@ezyang Fixed the problem in #11571
